### ISSUE TITLE
[SDK44] startLocationUpdatesAsync doesn't work when app is killed + Crashes App

### ISSUE THREAD

**Issue by islamouzou** [expo/expo#15515]
### Summary

i'm upgrading from SDK 42 to test the Background location which is already working fine but sometimes crashes with null pointer which is supposed to be fixed here https://github.com/expo/expo/pull/15023 , in changelogs i saw that's been fixed so i tried it on SDK44 . i'm building with the classic build system : expo build

Main Problem Description :  **startLocationUpdatesAsync doesn't work on killed state**

I start the bakground location , it works without a problem on foreground and background .
When i kill the app i'd like to have the location updates for my users who always kill their app .
Anyway : If i kill the App , i wait for while (10 - 20 min ) before the background location updates start , when it does , i see the foreground service but there are no location updates , values are not changing on my server . The same code on SDK42 Works with no problems . I don't know if it's an issue with location module or task-manager.

i attached a mini project to reproduce the Bug .

I hope you can fix this issue ASAP, it was already working fine with SDK42 except for this  issue #15023 
i have a big project based on background locations and i'm having a hard time with this .

Second problem : 

If i kill my app while the background location is on : i get this exception the next time i launch my app , it stucks on splash screen. (to catch this bug the location update must be started , so wait for the updates to start after you kill the app before you launch it again )

`java.lang.RuntimeException: Cannot get ReactInstanceManager from kernel
        at versioned.host.exp.exponent.ExpoTurboPackage.getDefaultUIManager(ExpoTurboPackage.kt:9)
        at versioned.host.exp.exponent.ExpoTurboPackage.getModule(ExpoTurboPackage.kt:3)
        at com.facebook.react.TurboReactPackage$ModuleHolderProvider.get(TurboReactPackage.java:2)
        at com.facebook.react.TurboReactPackage$ModuleHolderProvider.get(TurboReactPackage.java:1)
        at com.facebook.react.bridge.ModuleHolder.create(ModuleHolder.java:8)
        at com.facebook.react.bridge.ModuleHolder.getModule(ModuleHolder.java:7)
        at com.facebook.react.bridge.NativeModuleRegistry.getModule(NativeModuleRegistry.java:11)
        at com.facebook.react.bridge.CatalystInstanceImpl.getNativeModule(CatalystInstanceImpl.java:6)
        at com.facebook.react.bridge.CatalystInstanceImpl.getNativeModule(CatalystInstanceImpl.java:1)
        at com.facebook.react.bridge.ReactContext.getNativeModule(ReactContext.java:3)
        at versioned.host.exp.exponent.modules.api.reanimated.NodesManager.<init>(NodesManager.java:13)
        at versioned.host.exp.exponent.modules.api.reanimated.ReanimatedModule.getNodesManager(ReanimatedModule.java:2)
        at versioned.host.exp.exponent.modules.api.reanimated.ReanimatedJSIModulePackage.getJSIModules(ReanimatedJSIModulePackage.java:2)
        at versioned.host.exp.exponent.VersionedUtils$getReactInstanceManagerBuilder$builder$1.getJSIModules(VersionedUtils.kt:6)
        at com.facebook.react.ReactInstanceManager.createReactContext(ReactInstanceManager.java:21)
        at com.facebook.react.ReactInstanceManager.access$1100(ReactInstanceManager.java:1)
        at com.facebook.react.ReactInstanceManager$5.run(ReactInstanceManager.java:12)
        at java.lang.Thread.run(Thread.java:923)`

[backgroundLocationDemo.zip](https://github.com/expo/expo/files/7697687/backgroundLocationDemo.zip)

### Managed or bare workflow? If you have `ios/` or `android/` directories in your project, the answer is bare!

managed

### What platform(s) does this occur on?

Android

### SDK Version (managed workflow only)

44

### Environment

 Expo CLI 4.13.0 environment info:
    System:
      OS: Windows 10 10.0.19042
    Binaries:
      Node: 14.16.1 - C:\Program Files\nodejs\node.EXE
      Yarn: 1.22.10 - ~\AppData\Roaming\npm\yarn.CMD
      npm: 8.1.4 - C:\Program Files\nodejs\npm.CMD
    IDEs:
      Android Studio: Version     2020.3.0.0 AI-203.7717.56.2031.7583922
    npmPackages:
      expo: ^44.0.0-beta.1 => 44.0.0-beta.1
      react: 17.0.1 => 17.0.1
      react-dom: 17.0.1 => 17.0.1
      react-native: 0.64.3 => 0.64.3
      react-native-web: 0.17.1 => 0.17.1
    Expo Workflow: managed

### Reproducible demo

[backgroundLocationDemo.zip](https://github.com/expo/expo/files/7697687/backgroundLocationDemo.zip)

**Comment by islamouzou** [comment]
Un update that may be useful , the issue also appears on Expo Go , as you can see on this image , i start the background location on expo go , when i kill it , i can see the foreground service + location indicator , but there is no actual location updates .
![Screenshot_2021-12-12-01-15-24-395_com mi android globallauncher](https://user-images.githubusercontent.com/29413088/145695779-84e9ef8f-130e-401c-bf35-dc1f73134852.jpg)

**Comment by islamouzou** [comment]
Update 2 # The background location is working on killed state with EAS BUILD but not with expo build

**Comment by islamouzou** [comment]
Update 3 #


@cruzach @byCedric  @brentvatne 


I pretty much discovered the issue , it's react-native-reanimated 2.3.0 that is causing all what i mentionned  ( Stopping the background updates when app is killed + The Reanimated Exception obviously ) .

What i did ? 

i  switched to eas build , made a build with  react-native-reanimated 2.3.0 and another with 2.2.0

The results :

-The first build  2.3.0:  there is no location updates when app is killed .
-The seond build 2.2.0 :  Works like a charm , there are location updates as they are supposed to be .

( The two builds are made with "expo-task-manager": "~10.1.0" and "expo-location": "~14.0.0" ) So i'm pretty sure Reanimated is causing all of this .

 i hope it will be fixed ASAP before BETA Ends

**Comment by brentvatne** [comment]
@islamouzou - can you try building this project with EAS Build to see if you can reproduce the issue?

**Comment by islamouzou** [comment]
> @islamouzou - can you try building this project with EAS Build to see if you can reproduce the issue?

I'm on it

**Comment by islamouzou** [comment]
@brentvatne  i've created a new project using SDK44 , retested with expo build  just to be sure , i got the same behaviour .

EAS Build :  i gave it a try and i caugh this exception so far : 

`2021-12-13 22:10:49.235 32126-32155/? E/ReactNativeJS: 'TaskManager: Task "locationupdate" failed:', { [ReferenceError: Can't find variable: data] line: 466, column: 2293, sourceURL: 'index.android.bundle' }
2021-12-13 22:10:49.237 1643-2413/? E/JavaBinder: *** Uncaught remote exception!  (Exceptions are not yet supported across processes.)
    java.lang.RuntimeException: android.os.RemoteException: Couldn't get ApplicationInfo for package com.android.accelerometergraph
        at android.os.Parcel.writeException(Parcel.java:2161)
        at android.os.Binder.execTransactInternal(Binder.java:1178)
        at android.os.Binder.execTransact(Binder.java:1123)
     Caused by: android.os.RemoteException: Couldn't get ApplicationInfo for package com.android.accelerometergraph
        at com.android.server.pm.PackageManagerService$PackageManagerNative.getTargetSdkVersionForPackage(PackageManagerService.java:24513)
        at android.content.pm.IPackageManagerNative$Stub.onTransact(IPackageManagerNative.java:255)
        at android.os.Binder.execTransactInternal(Binder.java:1159)
        at android.os.Binder.execTransact(Binder.java:1123) 
2021-12-13 22:10:49.250 32126-32156/? E/unknown:ReactNative: console.error: TaskManager: Task "locationupdate" failed: ReferenceError: Can't find variable: data, stack:
    construct@-1
    p@109:351
    s@107:362
    construct@-1
    <unknown>@106:470
    v@106:658
    h@106:1835
    error@421:631
    <unknown>@477:3077
    y@123:586
    <unknown>@123:1891
    y@123:586
    o@123:1067
    <unknown>@123:1239
    f@119:154
    <unknown>@119:863
    k@125:535
    x@125:933
    callImmediates@125:3022
    value@50:2846
    <unknown>@50:1022
    value@50:2538
    value@50:992
    value@-1
    value@-1`

I'll keep digging to provide max useful data .

**Comment by islamouzou** [comment]
@brentvatne  i changed if (data) with if (locations) inside the task and rebuilt with eas build  now i don't  see that error but i'll check for other behaviours .

**Comment by islamouzou** [comment]
@brentvatne 

After many attempts to understand eas build behaviour here is what i got :

**The Very first launch after installation :** 

If i start the background location it works on background , foreground , if i **kill** it and wait for 1-2 minutes the foreground service reappear and the location updates start again .

Now if i start the app  again and Rekill it ,  the foreground service dissapears and location updates stop and never start . 

And no matter how many times i stop and start the location updates it will not work when app is Killed until i restart my phone or reinstall the app .

To sum up , the background location on Killed state only work for the first time after app installation or phone restart . 
I couldn't get the same Reanimated exeption as in expo build  but the background location updates in killed state are broken as well .

I hope this will lead somewhere .


[testApp.zip](https://github.com/expo/expo/files/7707629/testApp.zip)

**Comment by Kudo** [comment]
hi there! may i know how you kill the app? e.g. [in the reference](https://www.lifewire.com/close-apps-on-android-4164116), did you close apps by swiping up from home screen or force stopping from app settings?

**Comment by islamouzou** [comment]
@Kudo I kill it by swiping from home screen, in my case left or right on Redmi and Poco phones .

**Comment by Kudo** [comment]
thanks. shared my findings today.

1. there's a js error from your example where `data` is not defined.

```diff
-  async ({ data: { locations }, error }) => {
+  async ({ data, error }) => {
+    const { locations } = data;
```

2. `java.lang.RuntimeException: Cannot get ReactInstanceManager from kernel`
this happens from launch occasionally and possibly to be a race condition bug introduced by me.
i'll follow up this tomorrow.

3. background location not update
from `adb logcat`, i can see the background task service still running. there's no location update to js and sometimes showing `Calling JS function after bridge has been destroyed`. i will compare this with sdk42 apps and see if that's from expo modules or react-native 0.64 upgrade.

**Comment by islamouzou** [comment]
> thanks. shared my findings today.
> 
> 1. there's a js error from your example where `data` is not defined.
> 
> ```diff
> -  async ({ data: { locations }, error }) => {
> +  async ({ data, error }) => {
> +    const { locations } = data;
> ```
> 
> 2. `java.lang.RuntimeException: Cannot get ReactInstanceManager from kernel`
>    this happens from launch occasionally and possibly to be a race condition bug introduced by me.
>    i'll follow up this tomorrow.
> 3. background location not update
>    from `adb logcat`, i can see the background task service still running. there's no location update to js and sometimes showing `Calling JS function after bridge has been destroyed`. i will compare this with sdk42 apps and see if that's from expo modules or react-native 0.64 upgrade.



> thanks. shared my findings today.
> 
> 1. there's a js error from your example where `data` is not defined.
> 
> ```diff
> -  async ({ data: { locations }, error }) => {
> +  async ({ data, error }) => {
> +    const { locations } = data;
> ```
> 
> 2. `java.lang.RuntimeException: Cannot get ReactInstanceManager from kernel`
>    this happens from launch occasionally and possibly to be a race condition bug introduced by me.
>    i'll follow up this tomorrow.
> 3. background location not update
>    from `adb logcat`, i can see the background task service still running. there's no location update to js and sometimes showing `Calling JS function after bridge has been destroyed`. i will compare this with sdk42 apps and see if that's from expo modules or react-native 0.64 upgrade.

@Kudo  That's awesome , i'm glad you could reproduce the bugs , SDK 42  is the most stable for me till now in background loactions except for a null pointer exception which is already fixed i think in SDK 44 , couldn't test it yet due to this .

Thank you so much and good luck ^^

**Comment by islamouzou** [comment]
@Kudo @brentvatne  @byCedric  

Just for clarification , i want to know why does location background task takes some time to start after killing the app ? why doesn't the same foreground service persist  ? is it the expected behaviour ? this is not specific to SDK 44 by the way , i'm getting the same thing on SDK42 , i always have to wait before it starts .
